README.md

  ![slideparallax](https://img.shields.io/pub/v/slideparallax.svg) ![enter image description here](https://badges.bar/slideparallax/likes) ![enter image description here](https://badges.bar/slideparallax/popularity) ![enter image description here](https://badges.bar/slideparallax/pub%20points) ![enter image description here](https://img.shields.io/github/stars/muhammadsaddamnur/slideparallax.svg?style=flat&logo=github&colorB=deeppink&label=stars)  ![enter image description here](https://img.shields.io/badge/license-MIT-purple.svg) 

# slideparallax

Scrollview parallax inspired by the Tokopedia & Bukalapak application.  

## Usage

```dart
SlideParallax(
    header: Image.network(
        'https://images.tokopedia.net/img/cache/240/PYbRbC/2021/8/6/d1c4b788-27f3-463d-b40c-d2b1a6c845df.png.webp?ect=4g',
        height: 300,
    ),
    height: 300,
    width: MediaQuery.of(context).size.width,
    // headerSpacing: 150,
    itemCount: 10,
    itemBuilder: (BuildContext context, int index) {
        return Padding(
                padding: const EdgeInsets.all(1),
                child: Container(
                width: 150,
                color: Colors.red,
            ),
        );
    },
    footer: ElevatedButton(
        onPressed: () {
                print('object');
            },
        child: Text('Lainnya')
    ),
),
```

## Demo
![slideparallax](https://raw.githubusercontent.com/muhammadsaddamnur/slideparallax/main/slideparallax.gif)



## Support Me 🙏

  

<a  href="https://www.buymeacoffee.com/saddamnur">![buymeacoffee.com/saddamnur](https://www.buymeacoffee.com/assets/img/guidelines/download-assets-sm-1.svg)</a>

  

<b> Bisa pake Gopay, Dana, Link Aja, OVO, dan QRIS loh gaes.... </b>

<b>klik gambar dibawah ya </b>

<a  href="https://saweria.co/saddamnur">![enter image description here](https://raw.githubusercontent.com/muhammadsaddamnur/float_bubble/main/saweria.png)</a>

CHANGELOG.md
# 0.0.2🐞
* On tap header

# 0.0.1🐞
* release

test/slideparallax_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:slideparallax/slideparallax.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//   });
// }


lib/slideparallax.dart
library slideparallax;

import 'package:flutter/material.dart';

class SlideParallax extends StatefulWidget {
  ///height of slideparallax container
  ///
  final double height;

  ///width of slideparallax container
  ///
  final double width;

  ///spacing from header to content, default value is 2.5 of MediaQuery.of(context).size.width
  ///
  final double? headerSpacing;

  ///trigger will called animation header, default value is 50
  ///
  final double trigger;

  ///header widget
  ///
  final Widget header;

  ///on tap header widget
  ///
  final Function()? onTapHeader;

  ///footer widget
  ///
  final Widget? footer;

  ///creates a scrollable, linear array of widgets that are created on demand.
  ///
  final Widget Function(BuildContext, int) itemBuilder;

  ///This constructor is appropriate for list views with a large (or infinite) number of
  ///children because the builder is called only for those children that are actually visible.
  ///
  final int itemCount;

  const SlideParallax(
      {Key? key,
      required this.height,
      required this.width,
      this.headerSpacing,
      this.trigger = 50,
      required this.header,
      this.onTapHeader,
      required this.itemBuilder,
      required this.itemCount,
      this.footer})
      : super(key: key);

  @override
  _SlideParallaxState createState() => _SlideParallaxState();
}

class _SlideParallaxState extends State<SlideParallax> {
  ScrollController controller = ScrollController();

  double currentTrigger = 0;

  @override
  void initState() {
    currentTrigger = widget.trigger;
    if (mounted) {
      controller.addListener(() {
        if (controller.position.pixels >= widget.trigger) {
          if (currentTrigger != 0) {
            setState(() {
              if (currentTrigger >= 1) currentTrigger = 0;
            });
          }
        } else if (controller.position.pixels <= widget.trigger) {
          if (currentTrigger == 0) {
            setState(() {
              currentTrigger = widget.trigger;
            });
          }
        }
      });
    }
    super.initState();
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: widget.height,
      width: widget.width,
      child: Stack(
        children: [
          AnimatedPositioned(
            left: currentTrigger == widget.trigger ? 0 : widget.trigger * -1,
            duration: const Duration(milliseconds: 500),
            curve: Curves.fastOutSlowIn,
            child: AnimatedOpacity(
                opacity: currentTrigger == widget.trigger ? 1 : 0.5,
                duration: const Duration(milliseconds: 500),
                child: SizedBox(child: widget.header)),
          ),
          ScrollConfiguration(
              behavior: MyBehavior(),
              child: CustomScrollView(
                controller: controller,
                physics: PageScrollPhysics(),
                scrollDirection: Axis.horizontal,
                slivers: [
                  SliverToBoxAdapter(
                    child: Material(
                      color: Colors.transparent,
                      child: InkWell(
                        onTap: widget.onTapHeader,
                        child: Padding(
                          padding: const EdgeInsets.all(1),
                          child: SizedBox(
                            width: widget.headerSpacing == null
                                ? MediaQuery.of(context).size.width / 2.5
                                : widget.headerSpacing,
                          ),
                        ),
                      ),
                    ),
                  ),
                  SliverList(
                    delegate: SliverChildBuilderDelegate(widget.itemBuilder,
                        childCount: widget.itemCount),
                  ),
                  SliverToBoxAdapter(child: widget.footer ?? SizedBox()),
                ],
              )),
        ],
      ),
    );
  }
}

class MyBehavior extends ScrollBehavior {
  @override
  Widget buildViewportChrome(
      BuildContext context, Widget child, AxisDirection axisDirection) {
    return child;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:slideparallax/slideparallax.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Column(
          children: [
            SlideParallax(
              header: Image.network(
                'https://images.tokopedia.net/img/cache/240/PYbRbC/2021/8/6/d1c4b788-27f3-463d-b40c-d2b1a6c845df.png.webp?ect=4g',
                height: 300,
              ),
              onTapHeader: () {
                print('wkwk');
              },
              height: 300,
              width: MediaQuery.of(context).size.width,
              // headerSpacing: 150,
              itemCount: 10,
              itemBuilder: (BuildContext context, int index) {
                return Padding(
                  padding: const EdgeInsets.all(1),
                  child: Container(
                    width: 150,
                    color: Colors.red,
                  ),
                );
              },
              footer: ElevatedButton(
                  onPressed: () {
                    print('object');
                  },
                  child: Text('Lainnya')),
            ),
          ],
        ));
  }
}
